quesma/schema/hierarchical_schema.go

```diff
@@ -0,0 +1,47 @@
+// Copyright Quesma, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+package schema
+
+type SchemaTreeNode struct {
+	Name string // for example: "products" or "manufacturer" for "products.manufacturer", "" for root
+
+	// Either this is a leaf node (non-nil field) or a branch node (non-empty children)
+	Field    *Field
+	Children []*SchemaTreeNode
+}
+
+func addToTree(root *SchemaTreeNode, field *Field) {
+	components := field.PropertyName.Components()
+
+	leafComponent := components[len(components)-1]
+	components = components[:len(components)-1]
+
+	currentNode := root
+	for _, component := range components {
+		found := false
+		for _, child := range currentNode.Children {
+			if child.Name == component {
+				currentNode = child
+				found = true
+				break
```

Comment:
I'm debating whether removing `found`, namping COMPONENT_LOOP and using continue COMPONENT_LOOP would make more sense here.

Reply:
It would look like this:
```go
COMPONENT_LOOP:
	for _, component := range components {
		for _, child := range currentNode.Children {
			if child.Name == component {
				currentNode = child
				continue COMPONENT_LOOP
			}
		}

		newNode := SchemaTreeNode{Name: component}
		currentNode.Children = append(currentNode.Children, &newNode)
		currentNode = &newNode
	}
```

In that approach I don't like that at first sight those three lines:
```
		newNode := SchemaTreeNode{Name: component}
		currentNode.Children = append(currentNode.Children, &newNode)
		currentNode = &newNode
```
seem like they are executed in all iterations unconditionally, only at longer inspection a reader figures it out (that it's skipped with `continue`).

Therefore, I'm leaving this as-is for now.